qiime2/sdk/action.py

```diff
@@ -219,8 +255,16 @@ def bound_callable(*args, **kwargs):
 
                 # Wrap in a Results object mapping output name to value so
                 # users have access to outputs by name or position.
-                return qiime2.sdk.Results(self.signature.outputs.keys(),
-                                          outputs)
+                results = qiime2.sdk.Results(
+                    self.signature.outputs.keys(), outputs)
+
+                # If we are running a pipeline through parsl, we need to create
+                # a future here because the parsl join app the pipeline was
+                # running in is expected to return a future
+                if isinstance(self, Pipeline) and ctx.parsl:
+                    return _create_future(results)
```

Comment:
Let's move this `_create_future` up to `run_parsl_action` just before returning, since this only matters to parsl.